### PR TITLE
[CIS-166] Add Logger and related components

### PR DESCRIPTION
**Pull request by b-onc**
I still cannot decide how we're going to pass the logger around:
1. A global variable (not constant) that every new client will assign?
2. Dependency injecting it into every component?

### CODE REVIEW

**Comment by buh** [comment]
We can make as in Realm do. You can create the default configuration and use it with Realm instances as `let realm = Realm()` 
or `let realm = Realm(configuration: customConfiguration)`.
In our case: `let client = Client(apiKey: "...")` will use the default configuration with logger and etc.

**Comment by VojtaStavik** [comment]
@b-onc It would be really painful and impractical to pass the logger instance around. I'm completely OK with having one singleton instance of a logger in the module. The same way you have just one console output :)

**Comment by b-onc** [comment]
@buh That's already the case. I'm talking about passing logger around (ie using it in WebSocket, APIClient, controllers...)

**Comment by buh** [comment]
I think this is all about dependency injections. We can make an exception for the Logger, but then could be something else. The Client depends on other controllers that probably can have their own dependencies.

File: StreamChatClient_v3/Utils/Logger/Logger.swift [hunk 1]
```diff
@@ -0,0 +1,212 @@
+//
+// Logger.swift
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public var log: Logger {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Utils/Logger/Logger.swift:8]: nit: Missing comment

File: StreamChatClient_v3/Utils/Logger/Logger.swift [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  LogConfig.logger
+}
+
+public struct LogConfig {
+  /// Identifier for the logger. Defaults to empty.
+  public static var identifier = ""
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Utils/Logger/Logger.swift:14]: nit: If it defaults to empty, maybe it should rather be `String? = nil`

File: StreamChatClient_v3/Utils/Logger/Logger.swift [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  public static var showFileName = true
+
+  /// Toggle for showing line number in logs
+  public static var showLineNumber = true
+
+  /// Toggle for showing function name in logs
+  public static var showFunctionName = true
+
+  /// Destinations for the default logger. Please see `LogDestination`.
+  /// Defaults to only `ConsoleLogDestination`, which only prints the messages.
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Utils/Logger/Logger.swift:51]: nit: `only` ... `only`

File: StreamChatClient_v3/Utils/Logger/Logger.swift [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  ///
+  /// - Important: Other options in `ChatClientConfig.Logging` will not take affect if this is changed.
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Utils/Logger/Logger.swift:53]: outdated comment

File: StreamChatClient_v3/Utils/Logger/Logger.swift [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      showFileName: showFileName,
+      showLineNumber: showLineNumber,
+      showFunctionName: showFunctionName
+    )
+    return [consoleLogDestination]
+  }()
+
+  /// Logger instance to be used by StreamChat.
+  ///
+  /// - Important: Other options in `ChatClientConfig.Logging` will not take affect if this is changed.
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Utils/Logger/Logger.swift:72]: outdated

File: StreamChat.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -28,10 +28,10 @@
...
 		79280F542485529500CDEB89 /* ChannelEventsHandler_tests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F522485528800CDEB89 /* ChannelEventsHandler_tests.swift */; };
-		79280F782489181200CDEB89 /* URLSessionWebSocketEngine.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F772489181200CDEB89 /* URLSessionWebSocketEngine.swift */; };
-		79280F7A248918FD00CDEB89 /* StarscreamWebSocketEngine.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F79248918FD00CDEB89 /* StarscreamWebSocketEngine.swift */; };
 		79280F712487CD2B00CDEB89 /* Atomic.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F6F2487CD2B00CDEB89 /* Atomic.swift */; };
 		79280F732487CD3100CDEB89 /* Atomic_tests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F702487CD2B00CDEB89 /* Atomic_tests.swift */; };
+		79280F782489181200CDEB89 /* URLSessionWebSocketEngine.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F772489181200CDEB89 /* URLSessionWebSocketEngine.swift */; };
+		79280F7A248918FD00CDEB89 /* StarscreamWebSocketEngine.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79280F79248918FD00CDEB89 /* StarscreamWebSocketEngine.swift */; };
```
[the remark this answers is not in the document]
Comment (b-onc) [reply StreamChat.xcodeproj/project.pbxproj:34]: I totally agree! Somehow just adding new files generate these changes (everytime) so I thought it's best to include them in some PR, else everybody will have to discard these changes everytime

File: StreamChatClient_v3/Utils/Logger/Logger.swift (hunk 1 above)
Comment (b-onc) [reply StreamChatClient_v3/Utils/Logger/Logger.swift:8]: On a second thought, shouldn't this be `internal`? We can have internal loggers for all our SDKs, doesn't make sense to have a single public one

File: StreamChatClient_v3/Utils/Logger/Logger.swift (hunk 2 above)
Comment (b-onc) [reply StreamChatClient_v3/Utils/Logger/Logger.swift:14]: Gave it a thought, it's way easier to handle if it's empty

**Pull request merged**